platforms/software/maven/src/main/java/org/gradle/api/publish/maven/internal/artifact/AbstractMavenArtifact.java

```diff
@@ -28,43 +31,42 @@
 public abstract class AbstractMavenArtifact implements MavenArtifact, PublicationArtifactInternal {
     private final TaskDependency allBuildDependencies;
     private final DefaultTaskDependency additionalBuildDependencies;
-    private String extension;
-    private String classifier;
-
-    protected AbstractMavenArtifact(TaskDependencyFactory taskDependencyFactory) {
+    private final Property<String> extensionProperty;
+    private final Property<String> classifierProperty;
+
+    protected AbstractMavenArtifact(
+        TaskDependencyFactory taskDependencyFactory,
+        ObjectFactory objectFactory,
+        ProviderFactory providerFactory
+    ) {
         this.additionalBuildDependencies = new DefaultTaskDependency();
         this.allBuildDependencies = taskDependencyFactory.visitingDependencies(context -> {
             context.add(getDefaultBuildDependencies());
             additionalBuildDependencies.visitDependencies(context);
         });
+        this.extensionProperty = objectFactory.property(String.class);
+        this.classifierProperty = objectFactory.property(String.class);
+        // those should be lazy because the fields are not yet initialized in child classes
+        getExtension().set(providerFactory.provider(() -> getDefaultExtension().getOrNull()));
```

Comment:
❌ We should just do:
```
getExtension().set(getDefaultExtension())
```

Reply:
getDefaultExtension() is defined in a child class. It might be uninitialized.

Reply:
For example, `getDefaultExtension()` in `ArchiveTaskBasedMavenArtifact` returns `archiveTask.getArchiveExtension()` and the `archiveTask` field is not yet initialized

Reply:
I see. Yeah, I wonder if we should then init `defaultExtension` property in `AbstractMavenArtifact.java` and in subclasses we would assign a value to it, e.g.:
```
public ArchiveTaskBasedMavenArtifact() {
    super()
     getDefaultExtension().set(archiveTask.getArchiveExtension());
     getDefaultExtension().disallowChanges();
}
```

While `getExtension().set(providerFactory.provider(() -> getDefaultExtension().getOrNull()));` can work too, it looses provenance information

Reply:
I think this is fine for now. I think we should evaluate whether we even need special subclasses for all these types of artifacts, or if we could just have one implementation and each of these subclasses are replaced with instances of that one impl + some `convention` calls